### PR TITLE
Updates to shuffle_tfrecords_beam script for SparkRunner

### PR DESCRIPTION
**Pull request by anands-repo**
I understand that PRs are not performed on github. So, I just wanted to recommend/discuss some potential changes for the shuffle_tfrecords_beam.py script to enable running it with SparkRunner (PortableRunner). Without these changes the script works only in LOOPBACK mode (which is a testing mode where the actual work is performed on the submitting host).

### CODE REVIEW

**Comment by anands-repo** [comment]
@googlebot I signed it!

File: tools/shuffle_tfrecords_beam.py [hunk 1]
```diff
@@ -159,12 +159,21 @@ def sha1(input_bytes):
...
-  num_examples: {}
-  """.format(name, tfrecord_path, num_examples))
+# SparkRunner has some issue with using lambda functions in beam.Map
+# Hence, this has been turned into a callable
+class MakeConfigStringCaller:
+  def __init__(self, dataset_name, output_pattern_prefix):
+    self.name = dataset_name
+    self.tfrecord_path = output_pattern_prefix
+
+  def __call__(self, num_examples):
```
Comment (gunjanbaid) [inline tools/shuffle_tfrecords_beam.py:169]: The approach I've usually seen is defining a class that inherits from beam.DoFn. For example:

```
class MakeConfigStringCaller(beam.DoFn):

  def __init__(self, ...):
     ...

  def process(self, num_examples):
    ...
```

Usage:

```
config_str = (
  num_examples 
  | 'MakeConfigStr' >> beam.ParDo(MakeConfigStringCaller(dataset_name, output_pattern_prefix)))
```

Any reason to not use this approach? It's pretty similar to what you have, but is a Beam pattern I've seen used more commonly than a callable. I'm not sure which of these approaches would be faster though.

**Comment by gunjanbaid** [comment]
@anands-repo thanks for the pull request, we really appreciate community contributions! I added a comment about the approach used. 

You're right that we cannot merge PRs on GitHub directly, but once the code is finalized, we can review/test internally. If everything looks good and you are ok with it, we would submit the patch first to the internal codebase. The changes would be pushed to GitHub in the next release, and we would credit you in the commit description and release notes.

**Comment by anands-repo** [comment]
Hi @gunjanbaid , I am a novice when it comes to beam. So I will defer to you on https://github.com/google/deepvariant/pull/365#discussion_r512315819. I can try this out on my end as well. My setup works with Spark/Flink so I can test it out.

I am looking at another part of the code that is potentially fatal for execution at the moment:
```  
return (input_examples
          | 'Randomize' >> beam.Map(lambda x: (sha1(x), x))
          | 'Groupby' >> beam.GroupByKey()
          | 'DropKey' >> beam.FlatMap(lambda x: x[1]))
```

I notice that GroupByKey is loading all of the data into the memory of the worker. And I found corroboration for this (https://www.linkedin.com/pulse/spark-groupbys-unscalable-big-data-problem-bowen-gong). Is this not a problem for DataflowRunner?

I am trying to run shuffle on tfrecords produced from 6 BAM files. The gzipped tfrecords are approximately 120GB in total, and GroupByKey quickly runs out of memory when the machine has over 600GB of RAM.

**Comment by gunjanbaid** [comment]
@anands-repo my understanding of GroupByKey is what you have observed: all the data will be loaded into memory. If you are not able to use additional workers / use a larger machine with more memory, a workaround could be to run multiple shuffle jobs, each for a smaller subset of the data, rather than one global shuffle. I would also suggest contacting [Beam support](https://github.com/apache/beam#contact-us) to see if they can suggest any further optimization of this step.

**Comment by anands-repo** [comment]
Thanks for your comments @gunjanbaid 

I think, since everything else works well on single node systems, having the shuffle script work well on single node systems is also desirable.

It is good to be able to shuffle only smaller files like you said. But if we limit ourselves to the original tfrecord outputs, that comes with the limitation that the shuffles are localized and not global as in the current script.

However, a way to shuffle globally can be constructed from this idea with an additional step. This additional step will simply partition the input data into random buckets. Then we shuffle each bucket. I believe this is equivalent to a global shuffle with uniform probability for each permutation.

This would be something like:
```
input_data = readers | "FlattenInputs" >> beam.Flatten()
partitions = input_data | "PartitionInputs" >> beam.Partition(<random_partition_function_name>, <num_partitions>)
for i, p in enumerate(partitions):
    writing = p | "WritePartition%d" % i >> beam.io.WriteTFRecord(...)
```

Then each partition may be shuffled individually using the shuffle script.

I have rolled both partitioning and shuffling into the same [script](https://github.com/anands-repo/deepvariant/blob/r1.0/tools/shuffle_tfrecords_beam_for_local.py). I will report back regarding whether this works as expected. This depends on beam.Partition behaving properly.

**Comment by pichuan** [comment]
Thanks @anands-repo for the PR.
FYI: In the short term, we likely won't have bandwidth to confirm that this works. I'll leave this open and open an internal issue to track it. If/when we have a chance to confirm it works and merge it internally, we'll let you know (and will attribute it to your PR).

**Comment by anands-repo** [comment]
Hi @pichuan @gunjanbaid 

I apologize for overcomplicating this PR. I should have taken the rest of the discussion elsewhere. Thanks for your patient responses and advise so far.

Even though I made many commits here, the only PR here is for `shuffle_tfrecords_beam.py`, and the changes are trivial as @gunjanbaid has reviewed already. The purpose of the PR is to enable `shuffle_tfrecords_beam.py` to work with DirectRunner with multiple workers (`--direct_num_workers=2, --direct_running_mode="multi_processing"` for example), as well as with SparkRunner, FlinkRunner etc.

To answer @gunjanbaid 's question on using beam.DoFn instead of a callable, my personal opinion is that that may not be necessary. The issue I saw is that `lambda` functions cannot be invoked through ParDo for these runners/modes. I attribute it to the same issue that we see with python multiprocessing which uses pickle to dispatch functions across processes and lambda functions are not picklable. So in this context, I think a callable meets the minimum requirements to enable this.

I have tested on my end that the original deepvariant script and the modified script give the same output for a testcase.

I am not trying to push you into accepting this PR, but just trying to clear any confusions that may have been caused by my multiple commits to the same branch. I understand that the changes may still be far outside the priority of the team, so please take this at the appropriate pace.

Thanks!

**Pull request closed without merging**